language/move-prover/src/prelude.bpl

```diff
@@ -781,11 +796,16 @@ function {:inline} $Vector_type_value(tv: TypeValue): TypeValue {
 }
 
 function {:inline} $Vector_is_well_formed(v: Value): bool {
-    is#Vector(v) && l#ValueArray(v#Vector(v)) >= 0
+    is#Vector(v) && $vlen(v) >= 0 &&
+    (
+        var va := v#Vector(v);
+        0 <= l#ValueArray(va) &&
+        (forall x: int :: (0 <= x && x < l#ValueArray(va)) || v#ValueArray(va)[x] == DefaultValue)
```

Comment:
Why do we do this? We should never look at values in a ValueArray outside of index range. That is why we do have stratified IsEqual and other functions.

@shazqadeer

Reply:
It would be great if we maintained the invariant that all invalid elements of the vector are DefaultValue, even if we're not using it for ==.

However, if someone slips up and leaves a valid value off the end of a vector, could this result in inconsistencies?  Maybe there is a hazard in include the DefaultValue bit in Vector_is_well_formed.

Reply:
The last conjunct would be needed if we ever tried to make IsEqual an alias for ==. The array theory won't do the expected thing unless we have that conjunct.  I justified this conjunct to myself by arguing that all vectors constructed in our model go through our ValueArray constuctors.  If our constructors make sure that whenever each input satisfied that predicate, the output does also, then we have a strong reason to believe that no vector ever constructed will violate this predicate.  So, that could provide justification for including it in is_well_formed.

Reply:
While I agree this is true for (unspecified) values inside of a vector range, I'm not sure what the value outside of it is. When we display models in boogie_wrapper, we never look at values outside of the range. Is this for looking at raw models?

The disadvantages of something like this is that bugs can easily hide.  For example, the recent one with IsEqual vs ==. Perhaps == works in some cases with treatments like this but only until it does not. Also, reading the prelude may confuse people, because they get the impression this is required.